network/dag/tree/tree.go

```diff
@@ -0,0 +1,313 @@
+// If the path/leaf does not exist it will be created.
+func (t *tree) applyToPath(clock uint32, fn func(n *node) error) error {
+	// grow tree if needed
+	for clock >= t.MaxSize {
+		t.reRoot()
+	}
+
+	// apply fn to all nodes on path
+	var current *node
+	next := t.Root
+	for next != nil {
+		current = next
+		err := fn(current)
+		if err != nil {
+			return fmt.Errorf("failed for node with splitLC %d: %w", next.SplitLC, err)
+		}
+		next = t.getNextNode(current, clock)
+	}
+	t.dirtyLeaves[current.SplitLC] = current
+
+	return nil
+}
+
+func (t *tree) newBranch(start, stop uint32) *node {
+	split := (stop + start) / 2
+	n := newNode(split, stop, t.prototype.New())
+	if stop-start > t.LeafSize {
+		n.Left = t.newBranch(start, split)
+	} else {
+		t.dirtyLeaves[n.SplitLC] = n
+	}
+	return n
+}
+
+// reRoot creates a new Root with Data from the current Root and adds current Root as its Left branch.
+func (t *tree) reRoot() {
+	newRoot := newNode(t.MaxSize, 2*t.MaxSize, t.Root.Data.Clone())
+	newRoot.Left = t.Root
+	t.Root = newRoot
+	t.MaxSize *= 2
+	t.Depth++
+}
+
+// getNextNode retrieves the next node based on the clock value. If the node does not exist it is created.
+func (t *tree) getNextNode(n *node, clock uint32) *node {
+	// return nil if n is a leaf
+	if n.isLeaf() {
+		return nil
+	}
+
+	if clock < n.SplitLC {
+		return n.Left
+	} else {
+		if n.Right == nil {
+			n.Right = t.newBranch(n.SplitLC, n.LimitLC)
+		}
+		return n.Right
+	}
+}
+
+func (t *tree) GetRoot() Data {
+	return t.Root.Data.Clone()
+}
+
+func (t *tree) GetZeroTo(clock uint32) (Data, uint32) {
+	data := t.Root.Data.Clone()
+	next := t.Root
+	for {
+		current := next
+		if clock < current.SplitLC {
+			if current.Right != nil {
+				// Only fails when Data structures do not match, which should not happen for Data managed by the tree.
+				_ = data.Subtract(current.Right.Data)
+			}
+			next = current.Left
+		} else {
+			next = current.Right
+		}
+		if next == nil {
+			return data, rightmostLeafClock(current)
+		}
+	}
+}
+
+// rightmostLeafClock finds the rightmost leaf of the given node and returns its maximum clock value.
+func rightmostLeafClock(n *node) uint32 {
+	for {
+		if n.Right != nil {
+			n = n.Right
+		} else if n.Left != nil {
+			n = n.Left
+		} else {
+			return n.LimitLC - 1
+		}
+	}
+}
+
+func (t tree) GetUpdate() (dirty map[uint32][]byte, orphaned []uint32, err error) {
+	dirty = make(map[uint32][]byte, len(t.dirtyLeaves))
+	for k, v := range t.dirtyLeaves {
+		b, err := v.Data.MarshalBinary()
+		if err != nil {
+			return nil, nil, err
+		}
+		dirty[k] = b
+	}
+	for k := range t.orphanedLeaves {
+		orphaned = append(orphaned, k)
+	}
+	return dirty, orphaned, nil
+}
+
+func (t *tree) ResetUpdate() {
+	t.dirtyLeaves = map[uint32]*node{}
+	t.orphanedLeaves = nil
+}
+
+type dropLeavesUpdate struct {
+	dirty    map[uint32]*node
+	orphaned map[uint32]struct{}
+}
+
+func (t *tree) DropLeaves() {
+	// don't drop root
+	if t.Root == nil || t.Root.isLeaf() {
```

Comment:
if we only store actual data on the levaes, shouldn't we copy it to the parents?

Reply:
Data is stored on all nodes an parents contain the sum of their children data